docs/decisions/0004-error-handling.md

```diff
@@ -0,0 +1,59 @@
+
+- **Exception hierarchy inconsistency**: Refactor SK custom exceptions that are not derived from SKException to derive from SKException. This ensures consistency in the SK exception hierarchy.
+
+- **Missing original exception details**: Identify all cases where the original exception or failure is not preserved either as an inner exception or as a property/message of the rethrown exception, and address them. For example, AI service connectors that use HttClient should preserve the original status code by either setting it as an inner exception or including it as part of the rethrown exception's property.
+
+- **No consolidated way for exception handling**:
+  - [Memory connectors]: Introduce a new exception called VectorDbStorageException, along with derived exceptions like UpsertRecordException and RemoveRecordException, if additional details need to be captured and propagated to SK client code. Refactor all SK memory component implementations to use these exceptions instead of implementation-specific memory storage exceptions.
+  - [AI service connectors]: Restrict the usage of AIException to AI service connectors only and replace its usage in other SK components with more relevant exceptions. For cases where AIException is thrown, use HttpOperationException describved below as an inner exception to allow SK client code to access the original status code and exception/error details, if necessary.
+  - [HttpOperationException]: Create a new exception called HttpOperationException, which includes a StatusCode property, and implement the necessary logic to map the exception from HttpStatusCode, HttpRequestException, or Azure.RequestFailedException. Update existing SK code that interacts with the HTTP stack to throw HttpOperationException in case of a failed HTTP request and assign the original exception as its inner exception.
+
+- **Other**:
+  - Remove any code that wraps unhandled exceptions into AIException or any other SK exception solely for the purpose of wrapping. In most cases, this code does not provide useful information to help SK client code handle the exception, apart from a generic and uninformative "Something went wrong" message. Regardless of whether the unhandled exception is wrapped or not, SK client code cannot handle or recover from it without prior knowledge of the exception and proper exception handling logic.
+
+## Questions and consideration
+
+- Should the exception hierarchy follow the SK components hierarchy? For example, should there be a PlanningException for planners and a MemoryStorageException for memory storage connectors? What are the benefits for SK client code? Can the SK client code operate on the exception?  
```

Comment:
This comes down to what does the consuming code need to do.  What are the scenarios where consuming code needs to differentiate an exception related to planning from an exception related to memory storage? e.g. what are the scenarios where code would do:
```C#
try
{
   ...
}
catch (PlanningException pe)
{
     ... // do one thing
}
catch (MemoryStorageException mse)
{
    ... // do something else
}
```
If the consuming code would benefit from doing something different programmatically based on the exception type, differentiating can be valuable.  If not, differentiating adds little value and only cost.  One of my most costly mistakes for the .NET ecosystem was introducing a particular exception type when I shouldn't have.

Reply:
Be honest, at the moment, I don't see any scenario where consuming code needs to differentiate exceptions between different SK components.